judge/judgedaemon.main.php

```diff
@@ -133,32 +137,49 @@ function request(string $url, string $verb = 'GET', $data = '', bool $failonerro
         curl_setopt($curl_handle, CURLOPT_POSTFIELDS, null);
     }
 
-    $response = curl_exec($curl_handle);
-    if ($response === false) {
-        $errstr = "Error while executing curl $verb to url " . $url . ": " . curl_error($curl_handle);
-        if ($failonerror) {
-            error($errstr);
+    $delay_in_ms = 1000;
+    $succeeded = false;
+    $response = null;
+    $errstr = null;
+
+    for ($trial = 1; $trial <= BACKOFF_STEPS; $trial++) {
+        $response = curl_exec($curl_handle);
+        if ($response === false) {
+            $errstr = "Error while executing curl $verb to url " . $url . ": " . curl_error($curl_handle);
         } else {
-            warning($errstr);
-            $endpoints[$endpointID]['errorred'] = true;
-            return null;
+            $status = curl_getinfo($curl_handle, CURLINFO_HTTP_CODE);
+            if ($status == 401) {
+                $errstr = "Authentication failed (error $status) while contacting $url. " .
+                    "Check credentials in restapi.secret.";
+                break;
+            } else if ($status < 200 || $status >= 300) {
+                $json = dj_json_try_decode($response);
+                if ($json !== null) {
+                    $response = var_export($json, true);
+                }
+                $errstr = "Error while executing curl $verb to url " . $url .
+                    ": http status code: " . $status .
+                    ", request size = " . strlen(print_r($data, true)) .
+                    ", response: " . $response;
+                if ($status == 500) {
+                    break;
+                }
+            } else {
+                $succeeded = true;
+                break;
+            }
         }
-    }
-    $status = curl_getinfo($curl_handle, CURLINFO_HTTP_CODE);
-    if ($status < 200 || $status >= 300) {
-        if ($status == 401) {
-            $errstr = "Authentication failed (error $status) while contacting $url. " .
-                "Check credentials in restapi.secret.";
+        if ($trial == BACKOFF_STEPS) {
+            $errstr = $errstr . " Retry limit reached.";
         } else {
-            $json = dj_json_decode($response);
-            if ($json !== null) {
-                $response = var_export($json, true);
-            }
-            $errstr = "Error while executing curl $verb to url " . $url .
-                ": http status code: " . $status .
-                ", request size = " . strlen(print_r($data, true)) .
-                ", response: " . $response;
+            $warnstr = $errstr . " This request will be retried after about " .
+                $delay_in_ms . "ms... (" . $trial . "/" . BACKOFF_STEPS . ")";
+            warning($warnstr);
+            usleep($delay_in_ms + random_int(0, BACKOFF_JITTER_MS));
```

Comment:
This is now sleeping after the last unsuccessful attempt as well, isn't it?

Reply:
I don't think so. The function does not sleep after the last unsuccessful attempt because `$trial` is equal to `BACKOFF_STEPS` on the last repetition.

Reply:
You are right, I missed the condition above.